### PR TITLE
Postgers implementation and databases setup

### PR DESCRIPTION
**Pull request by IshmeetSingh06**
This branch consists of postures implementation and creates the tables if they don't exist

### CODE REVIEW

File: lib/library_db.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  def initialize
+    begin
+      self.conn = PG.connect(
+        dbname: 'postgres',
+        user: ENV['PG_DATABASE_USERNAME'],
+        password: ENV['PG_DATABASE_PASSWORD'],
+        port: ENV['PG_DATABASE_PORT']
+      )
+    rescue PG::Error => error
+      puts "An error occurred while connecting to postgres : #{error.message}"
```
Comment (neeraj33negi) [inline lib/library_db.rb]: ideally app should exit when DB connection fails,

File: lib/user.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class User
+  attr_reader :id
+  attr_accessor :username, :password, :first_name, :last_name, :admin
+
+  def initialize(username, password, first_name, last_name = nil, admin = false)
+    self.username= username
+    self.password= password
+    self.first_name= first_name
+    self.last_name= last_name unless last_name.nil? || last_name.empty?
```
Comment (neeraj33negi) [inline lib/user.rb]: space before `=`

Comment (neeraj33negi) [reply lib/user.rb]: also we dont need handling this, adding these conditions has no impact on what is assigned to last_name

File: lib/library_db.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
+class LibraryDB
```
Comment (gagan93jtg) [inline lib/library_db.rb]: can we call it `DatabaseConnector`? Our app might expand and just calling this `LibraryDB` might confuse.

File: lib/library_db.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      self.conn = PG.connect(
+        dbname: 'postgres',
+        user: ENV['PG_DATABASE_USERNAME'],
+        password: ENV['PG_DATABASE_PASSWORD'],
+        port: ENV['PG_DATABASE_PORT']
+      )
+    rescue PG::Error => error
+      puts "An error occurred while connecting to postgres :\n #{error.message}"
+      puts "Exiting the application.................."
+      exit(false)
```
Comment (gagan93jtg) [inline lib/library_db.rb]: See how we can use return codes to send some useful information to operating system (like `exit(1)` or something similar)

File: lib/library_db.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          return_time DATE
+        );
+      SQL
+    )
+  end
+
+  private def initialize_admin_account
+    return if admin_account_exists?
+
+    hashed_password = BCrypt::Password.create('admin@123')
```
Comment (gagan93jtg) [inline lib/library_db.rb]: can we take this from ENV so that we can configure our own admin password?

File: lib/library_db.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    @conn.exec_params(
+      'INSERT INTO users (username, password, first_name, admin) VALUES ($1, $2, $3, $4)',
+      ['admin@library', hashed_password, 'admin', true]
+    )
+  end
+
+  private def admin_account_exists?
+    result = self.conn.exec_params('SELECT EXISTS (SELECT 1 FROM users WHERE username = $1)',['admin@library'])
```
Comment (gagan93jtg) [inline lib/library_db.rb]: why nested query? can we write this without nesting?

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+require './lib/library_app.rb'
```
Comment (gagan93jtg) [inline main.rb]: Let's call this runner.rb or boot.rb. Such names are used for files that startup the application

File: lib/library_app.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class LibraryApp
```
Comment (gagan93jtg) [inline lib/library_app.rb]: `application.rb`

File: lib/book.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+class Book
```
Comment (gagan93jtg) [inline lib/book.rb]: Three PORO classes can be moved to a nested folder because they're different from rest of the files :)

**Review by gagan93jtg — CHANGES_REQUESTED** [review]

File: lib/database_connector.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  end
+
+  private def initialize_admin_account
+    return if admin_account_exists?
+
+    hashed_password = BCrypt::Password.create(ENV['PG_ADMIN_PASSWORD'])
+
+    @conn.exec_params(
+      'INSERT INTO users (username, password, first_name, admin) VALUES ($1, $2, $3, $4)',
+      ['admin@library', hashed_password, 'admin', true]
```
Comment (gagan93jtg) [inline lib/database_connector.rb]: 1. why hardcoding this when we have taken this out in the ENV variable (as checking the same in lint 79).
2. Let's not put `PG` prefix here. This environment configuration is not related to postgres. It's our admin user, not postgres admin user.

**Comment by PramanSatya** [comment]
Please change base branch

**Review by gagan93jtg — APPROVED** [review]
LGTM. Please merge after @PramanSatya's approval

**Comment by gagan93jtg** [comment]
@IshmeetSingh06 there are some conflicts from the base branch. Please resolve

**Review by PramanSatya — APPROVED** [review]

**Pull request merged**